File: programmers/old/level3/단속카메라.java
import java.util.*;
class Solution {
    public int solution(int[][] routes) {
        int answer = 0;
        Arrays.sort(routes, new Comparator<int[]>() {
        	public int compare(int[] a, int[] b) {
        		return Integer.compare(a[0],  b[0]);
        	}
		});
        PriorityQueue<int[]> pq = new PriorityQueue<int[]>(new Comparator<int[]>() {
        	public int compare(int[] a, int[] b) {
        		return Integer.compare(a[1],  b[1]);
        	}
        });
        int idx = 0;
        while(idx < routes.length) {
        	if(!pq.isEmpty()) {
	        	if(routes[idx][0] < pq.element()[1]) {
	        		pq.add(routes[idx++]);
	        	}else if(routes[idx][0] > pq.element()[1]) {
	        		answer++;
	        		while(!pq.isEmpty()) {
	        			pq.remove();
	        		}
	        	}else {
	        		while(routes[idx][0] == pq.element()[1]) {
	        			idx++;
	        		}
	        		answer++;
	        		while(!pq.isEmpty()) {
	        			pq.remove();
	        		}
	        	}
        	}else {
        		pq.add(routes[idx++]);
        	}
        }
        answer++;
        return answer;
    }
}
public class 단속카메라 {
	public static void main(String[] args) {
		Solution sol = new Solution();
		int[][] routes = {{-20,15}, {-14,-5}, {-18,-13}, {-5,-3}};
		System.out.println(sol.solution(routes));
	}
}

File: programmers/js/new/level3/단속카메라.js
function solution(routes) {
    var answer = 1;
    routes.sort((a, b) => {
        if (a[1] < b[1]) {
            return -1;
        } else if (a[1] > b[1]) {
            return 1;
        }
        return a[0] - b[0];
    });
    let arr = [routes[0]];
    for (let i = 1; i < routes.length; ++i) {
        const current = routes[i];
        if (arr.length > 0 && current[0] > arr[0][1]) {
            arr = [];
            answer++;
        }
        arr.push(current);
    }
    return answer;
}

console.log(
    solution([
        [-15, -14],
        [-14, 1],
        [2, 3],
    ])
);
